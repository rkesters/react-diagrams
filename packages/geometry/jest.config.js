const sharedConfig = require('../../jest.config.js')(__dirname);
module.exports = {
	...sharedConfig,
	testEnvironment: 'jsdom',

};
