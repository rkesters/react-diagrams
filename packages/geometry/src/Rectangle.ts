import { isNumber } from 'lodash';
import { Point } from './Point';
import { Polygon } from './Polygon';

export class Rectangle extends Polygon {
	constructor(tl: Point, tr: Point, br: Point, bl: Point);
	constructor(position: Point, width: number, height: number);
	constructor(x?: number, y?: number, width?: number, height?: number);

	constructor(a: Point | number = 0, b: Point | number  = 0, c: Point | number  = 0, d: Point | number  = 0) {
		if (Point.isa(a) && Point.isa(b) && Point.isa(c) && Point.isa(d)) {
			super([a, b, c, d]);
			return;
		} else if (Point.isa(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super([a, new Point(a.x + b, a.y), new Point(a.x + b, a.y + c), new Point(a.x, a.y + c)]);
			return;
		} else if (isNumber(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super(Rectangle.pointsFromBounds(a, b, c, d));
			return;
		}

		throw new Error(`Rectangle call with invaild arguements: ${arguments.length}`);
	}

	static pointsFromBounds(x: number, y: number, width: number, height: number): Point[] {
		return [new Point(x, y), new Point(x + width, y), new Point(x + width, y + height), new Point(x, y + height)];
	}

	updateDimensions(x: number, y: number, width: number, height: number) {
		this.points = Rectangle.pointsFromBounds(x, y, width, height);
	}

	setPoints(points: Point[]) {
		if (points.length !== 4) {
			throw 'Rectangles must always have 4 points';
		}
		super.setPoints(points);
	}

	containsPoint(point: Point) {
		const tl = this.getTopLeft();
		const br = this.getBottomRight();

		return point.x >= tl.x && point.x <= br.x && point.y >= tl.y && point.y <= br.y;
	}

	getWidth(): number {
		return Math.sqrt(
			Math.pow(this.getTopLeft().x - this.getTopRight().x, 2) + Math.pow(this.getTopLeft().y - this.getTopRight().y, 2)
		);
	}

	getHeight(): number {
		return Math.sqrt(
			Math.pow(this.getBottomLeft().x - this.getTopLeft().x, 2) +
				Math.pow(this.getBottomLeft().y - this.getTopLeft().y, 2)
		);
	}

	getTopMiddle(): Point {
		return Point.middlePoint(this.getTopLeft(), this.getTopRight());
	}

	getBottomMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getBottomRight());
	}

	getLeftMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getTopLeft());
	}

	getRightMiddle(): Point {
		return Point.middlePoint(this.getBottomRight(), this.getTopRight());
	}

	getTopLeft(): Point {
		return this.points[0];
	}

	getTopRight(): Point {
		return this.points[1];
	}

	getBottomRight(): Point {
		return this.points[2];
	}

	getBottomLeft(): Point {
		return this.points[3];
	}
}
