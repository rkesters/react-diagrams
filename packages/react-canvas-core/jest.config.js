const sharedConfig = require('../../jest.config.js')(__dirname);
module.exports = {
	...sharedConfig,
	testEnvironment: 'jsdom',
	collectCoverageFrom: [
		...sharedConfig.collectCoverageFrom,
		'!src/index.ts'
	],
};
