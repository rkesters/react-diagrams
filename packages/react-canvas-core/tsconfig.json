{
	"extends": "../../tsconfig.base.json",
	"compilerOptions": {
		"outDir": "dist",
		"rootDir": "src",
		"declarationDir": "dist/@types",
		"esModuleInterop": true
	},
	"include": [
		"./src"
	],
	"references": [
		{
			"path": "../geometry"
		}
	]
}
