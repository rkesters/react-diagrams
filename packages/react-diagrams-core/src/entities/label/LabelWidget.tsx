import * as React from 'react';
import { DiagramEngine } from '../../DiagramEngine';
import { LabelModel } from './LabelModel';
import styled from '@emotion/styled';
import { get } from 'lodash';

export interface LabelWidgetProps {
	engine: DiagramEngine;
	label: LabelModel;
	index: number;
}

namespace S {
	export const Label = styled.div`
		display: inline-block;
		position: absolute;
	`;

	export const Foreign = styled.foreignObject`
		pointer-events: none;
		overflow: visible;
	`;
}

export class LabelWidget extends React.Component<LabelWidgetProps> {
	ref: React.RefObject<HTMLDivElement>;

	constructor(props: LabelWidgetProps) {
		super(props);
		this.ref = React.createRef();
	}

	componentDidUpdate() {
		window.requestAnimationFrame(this.calculateLabelPosition);
	}

	componentDidMount() {
		window.requestAnimationFrame(this.calculateLabelPosition);
	}

	findPathAndRelativePositionToRenderLabel = (index: number): { path: SVGPathElement; position: number } | undefined => {
		// an array to hold all path lengths, making sure we hit the DOM only once to fetch this information
		const link = this.props.label.getParent();
		const paths =link?.getRenderedPath();
		if (!link || !paths) {
			throw new Error(`Failed to findPathAndRelativePositionToRenderLabel for index ${index}`);
		}
		const lengths = paths.map((path) => path.getTotalLength()) ?? [];

		// calculate the point where we want to display the label
		let labelPosition =
			lengths?.reduce((previousValue, currentValue) => previousValue + currentValue, 0) *
			(index / (link.getLabels().length + 1));

		// find the path where the label will be rendered and calculate the relative position
		let pathIndex = 0;
		while (pathIndex < paths.length) {
			if (labelPosition - lengths[pathIndex] < 0) {
				return {
					path: paths[pathIndex],
					position: labelPosition
				};
			}

			// keep searching
			labelPosition -= lengths[pathIndex];
			pathIndex++;
		}
	};

	calculateLabelPosition = () => {
		if (!this.ref.current) {
			return;
		}
		const found = this.findPathAndRelativePositionToRenderLabel(this.props.index + 1);
		if (!found) {
			return;
		}

		const { path, position } = found;

		const labelDimensions = {
			width: this.ref.current.offsetWidth ?? 0,
			height: this.ref.current.offsetHeight ?? 0
		};

		const pathCentre = path.getPointAtLength(position);

		const labelCoordinates = {
			x: pathCentre.x - labelDimensions.width / 2 + get(this.props.label.getOptions(), 'offsetX', 0),
			y: pathCentre.y - labelDimensions.height / 2 + get(this.props.label.getOptions(), 'offsetY', 0)
		};

		this.ref.current.style.transform = `translate(${labelCoordinates.x}px, ${labelCoordinates.y}px)`;
	};

	render() {
		const canvas = this.props.engine.getCanvas();

		return (
			<S.Foreign key={this.props.label.getID()} width={canvas?.offsetWidth} height={canvas?.offsetHeight}>
				<S.Label ref={this.ref}>
					{this.props.engine.getFactoryForLabel(this.props.label).generateReactWidget({ model: this.props.label })}
				</S.Label>
			</S.Foreign>
		);
	}
}
