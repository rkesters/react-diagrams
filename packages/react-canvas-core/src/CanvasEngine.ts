import { debounce } from 'lodash';
import { CanvasModel } from './entities/canvas/CanvasModel';
import { FactoryBank } from './core/FactoryBank';
import { AbstractReactFactory } from './core/AbstractReactFactory';
import { LayerModel } from './entities/layer/LayerModel';
import { BaseListener, BaseObserver } from './core/BaseObserver';
import { MouseEvent } from 'react';
import { BaseModel } from './core-models/BaseModel';
import { Point } from '@projectstorm/geometry';
import { ActionEventBus } from './core-actions/ActionEventBus';
import { ZoomCanvasAction } from './actions/ZoomCanvasAction';
import { DeleteItemsAction } from './actions/DeleteItemsAction';
import { StateMachine } from './core-state/StateMachine';
export interface CanvasEngineListener extends BaseListener {
	canvasReady(): void;

	repaintCanvas(): void;

	rendered(): void;
}

/**
 * Defines the CanvasEngine options
 */
export interface CanvasEngineOptions {
	registerDefaultDeleteItemsAction?: boolean;
	registerDefaultZoomCanvasAction?: boolean;
	/**
	 * Defines the debounce wait time in milliseconds if > 0
	 */
	repaintDebounceMs?: number;
}

export class CanvasEngine<
	L extends CanvasEngineListener = CanvasEngineListener,
	M extends CanvasModel = CanvasModel
> extends BaseObserver<L> {
	protected model: M ;
	protected layerFactories: FactoryBank<AbstractReactFactory<LayerModel>>;
	protected canvas: HTMLDivElement | undefined;
	protected eventBus: ActionEventBus;
	protected stateMachine: StateMachine;
	protected options: CanvasEngineOptions;

	constructor(model: M, options: CanvasEngineOptions = {}) {
		super();
		this.model = model;
		this.eventBus = new ActionEventBus(this);
		this.stateMachine = new StateMachine(this);
		this.layerFactories = new FactoryBank();
		this.registerFactoryBank(this.layerFactories);

		/**
		 * Overrides the standard options with the possible given options
		 */
		this.options = {
			registerDefaultDeleteItemsAction: true,
			registerDefaultZoomCanvasAction: true,
			repaintDebounceMs: 0,
			...options
		};
		if (this.options.registerDefaultZoomCanvasAction === true) {
			this.eventBus.registerAction(new ZoomCanvasAction({}, this));
		}
		if (this.options.registerDefaultDeleteItemsAction === true) {
			this.eventBus.registerAction(new DeleteItemsAction({}, this));
		}
	}

	getStateMachine() {
		return this.stateMachine;
	}

	getRelativeMousePoint(event: { clientX: number; clientY: number }): Point {
		const point = this.getRelativePoint(event.clientX, event.clientY);
		return new Point(
			(point.x - (this.model?.getOffsetX() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0),
			(point.y - (this.model?.getOffsetY() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0)
		);
	}

	getRelativePoint(x:number, y: number): Point {
		const canvasRect = this.canvas?.getBoundingClientRect();
		if (!canvasRect) {
			throw new Error(`Can't get relative point because canvas has not been defined.`)
		}
		return new Point(x - canvasRect.left, y - canvasRect.top);
	}

	registerFactoryBank(factory: FactoryBank<any>) {
		factory.registerListener({
			factoryAdded: (event) => {
				event.factory.setDiagramEngine(this);
			},
			factoryRemoved: (event) => {
				event.factory.setDiagramEngine(null);
			}, eventWillFire: ():void => {} , eventDidFire: ():void => {}
		});
	}

	getActionEventBus() {
		return this.eventBus;
	}

	getLayerFactories() {
		return this.layerFactories;
	}

	getFactoryForLayer<F extends AbstractReactFactory<LayerModel>>(layer: LayerModel | string | undefined) {
		if (!layer) { return;}
		if (typeof layer === 'string') {
			return this.layerFactories.getFactory(layer);
		}
		return this.layerFactories.getFactory(layer.getType());
	}

	setModel(model: M) {
		this.model = model;
		if (this.canvas) {
			requestAnimationFrame(() => {
				this.repaintCanvas();
			});
		}
	}

	getModel(): M {
		return this.model;
	}

	repaintCanvas(promise: true): Promise<any>;
	repaintCanvas(): void;
	repaintCanvas(promise?: true): Promise<any> | void {
		const repaintDebounceMs  = this.options.repaintDebounceMs ?? 0;

		/**
		 * The actual repaint function
		 */
		const repaint = () => {
			this.iterateListeners((listener) => {
				if (listener.repaintCanvas) {
					listener.repaintCanvas();
				}
			});
		};

		// if the `repaintDebounceMs` option is > 0, then apply the debounce
		let repaintFn = repaint;

		if (repaintDebounceMs > 0) {
			repaintFn = debounce(repaint, repaintDebounceMs);
		}

		if (promise) {
			return new Promise<void>((resolve) => {
				const l = this.registerListener({
					rendered: () => {
						resolve();
						l.deregister();
					}
				} as L);
				repaintFn();
			});
		}

		repaintFn();
	}

	setCanvas(canvas?: HTMLDivElement | null) {
		if (this.canvas !== canvas) {
			this.canvas = canvas ?? undefined;
			if (canvas) {
				this.fireEvent(undefined, 'canvasReady');
			}
		}
	}

	getCanvas() {
		return this.canvas;
	}

	getMouseElement(event: MouseEvent): BaseModel<any> | null | undefined  {
		return null;
	}

	zoomToFit() {
		if (!this.canvas) {
			throw new Error(`Can not zoom to fit because canvas is not defined`)
		}
		const xFactor = this.canvas.clientWidth / this.canvas.scrollWidth;
		const yFactor = this.canvas.clientHeight / this.canvas.scrollHeight;
		const zoomFactor = xFactor < yFactor ? xFactor : yFactor;

		this.model?.setZoomLevel((this.model?.getZoomLevel() ?? 1) * zoomFactor);
		this.model?.setOffset(0, 0);
		this.repaintCanvas();
	}
}
