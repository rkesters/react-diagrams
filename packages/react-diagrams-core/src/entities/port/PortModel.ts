import { NodeModel } from '../node/NodeModel';
import { LinkModel } from '../link/LinkModel';
import * as _ from 'lodash';
import { Point, Rectangle } from '@projectstorm/geometry';
import {
	BaseEntityEvent,
	BaseModelOptions,
	BasePositionModel,
	BasePositionModelGenerics,
	BasePositionModelListener,
	DeserializeEvent
} from '@projectstorm/react-canvas-core';

export enum PortModelAlignment {
	TOP = 'top',
	LEFT = 'left',
	BOTTOM = 'bottom',
	RIGHT = 'right'
}

export interface PortModelListener extends BasePositionModelListener {
	/**
	 * fires when it first receives positional information
	 */
	reportInitialPosition: (event: BaseEntityEvent<PortModel>) => void;
}

export interface PortModelOptions extends BaseModelOptions {
	alignment?: PortModelAlignment;
	maximumLinks?: number;
	name: string;
}

export interface PortModelGenerics extends BasePositionModelGenerics {
	OPTIONS: PortModelOptions;
	PARENT: NodeModel;
	LISTENER: PortModelListener;
}

export class PortModel<G extends PortModelGenerics = PortModelGenerics> extends BasePositionModel<G> {
	links: { [id: string]: LinkModel };

	// calculated post rendering so routing can be done correctly
	width: number;
	height: number;
	reportedPosition: boolean;

	constructor(options: G['OPTIONS']) {
		super(options);
		this.links = {};
		this.reportedPosition = false;
		this.width = 0;
		this.height = 0;
	}

	deserialize(event: DeserializeEvent<ReturnType<PortModel['serialize']>>) {
		super.deserialize(event);
		this.reportedPosition = false;
		this.options.name = event.data.name;
		this.options.alignment = event.data.alignment;
	}

	serialize() {
		return {
			...super.serialize(),
			name: this.options.name,
			alignment: this.options.alignment,
			parentNode: this.parent?.getID(),
			links: _.map(this.links, (link) => {
				return link.getID();
			})
		};
	}

	setPosition(point: Point): void;
	setPosition(x: number, y: number): void;
	setPosition(x: Point | number, y?: number) {
		let old = this.position;
		Point.isa(x) ? super.setPosition(x) : super.setPosition(x, y as number) ;
		const X = this.getPosition().x;
		const Y = this.getPosition().y;

		_.forEach(this.getLinks(), (link) => {
			let point = link.getPointForPort(this);
			point?.setPosition(point.getX() + X - old.x, point.getY() + Y - old.y);
		});
	}

	doClone(lookupTable = {}, clone: this) {
		clone.links = {};
		clone.parent = this.getParent()?.clone(lookupTable);
	}

	getNode(): NodeModel | undefined {
		return this.getParent();
	}

	getName(): string {
		return this.options.name;
	}

	getMaximumLinks(): number | undefined  {
		return this.options.maximumLinks;
	}

	setMaximumLinks(maximumLinks: number) {
		this.options.maximumLinks = maximumLinks;
	}

	removeLink(link: LinkModel) {
		delete this.links[link.getID()];
	}

	addLink(link: LinkModel) {
		this.links[link.getID()] = link;
	}

	getLinks(): { [id: string]: LinkModel } {
		return this.links;
	}

	public createLinkModel(): LinkModel | null {
		if (_.isFinite(this.options.maximumLinks)) {
			var numberOfLinks: number = _.size(this.links);
			if (this.options.maximumLinks === 1 && numberOfLinks >= 1) {
				return _.values(this.links)[0];
			} else if (numberOfLinks >= (this.options.maximumLinks ?? Number.MAX_SAFE_INTEGER/2)) {
				return null;
			}
		}
		return null;
	}

	reportPosition() {
		_.forEach(this.getLinks(), (link) => {
			link.getPointForPort(this)?.setPosition(this.getCenter());
		});
		this.fireEvent(
			{
				entity: this
			},
			'reportInitialPosition'
		);
	}

	getCenter(): Point {
		return new Point(this.getX() + this.width / 2, this.getY() + this.height / 2);
	}

	getBoundingBox(): Rectangle {
		return new Rectangle(this.position, this.width, this.height);
	}

	updateCoords(coords: Rectangle) {
		this.width = coords.getWidth();
		this.height = coords.getHeight();
		this.setPosition(coords.getTopLeft());
		this.reportedPosition = true;
		this.reportPosition();
	}

	canLinkToPort(port: PortModel): boolean {
		return true;
	}

	isLocked(): boolean {
		return super.isLocked() || (this.getParent()?.isLocked() ?? false);
	}
}
