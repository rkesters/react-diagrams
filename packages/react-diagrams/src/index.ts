import {
	DefaultDiagramState,
	DiagramEngine,
	DiagramModel,
	LinkLayerFactory,
	NodeLayerFactory
} from '@projectstorm/react-diagrams-core';
import {
	DefaultLabelFactory,
	DefaultLinkFactory,
	DefaultNodeFactory,
	DefaultPortFactory
} from '@projectstorm/react-diagrams-defaults';
import { PathFindingLinkFactory } from '@projectstorm/react-diagrams-routing';
import { SelectionBoxLayerFactory, CanvasEngineOptions } from '@projectstorm/react-canvas-core';

export * from '@projectstorm/react-diagrams-core';
export * from '@projectstorm/react-diagrams-defaults';
export * from '@projectstorm/react-diagrams-routing';

/**
 * Construct an engine with the defaults installed
 */
export default (model: DiagramModel, options: CanvasEngineOptions = {}): DiagramEngine => {
	const engine = new DiagramEngine(model,options);

	// register model factories
	engine.getLayerFactories().registerFactory(new NodeLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new LinkLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new SelectionBoxLayerFactory(engine));

	engine.getLabelFactories().registerFactory(new DefaultLabelFactory(engine));
	engine.getNodeFactories().registerFactory(new DefaultNodeFactory(engine)); // i cant figure out why
	engine.getLinkFactories().registerFactory(new DefaultLinkFactory(engine));
	engine.getLinkFactories().registerFactory(new PathFindingLinkFactory(engine));
	engine.getPortFactories().registerFactory(new DefaultPortFactory(engine));

	// register the default interaction behaviours
	engine.getStateMachine().pushState(new DefaultDiagramState(engine));
	return engine;
};
