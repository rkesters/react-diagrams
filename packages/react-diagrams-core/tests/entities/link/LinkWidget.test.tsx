import { Toolkit } from '@projectstorm/react-canvas-core';
import { render } from '@testing-library/react';
import React from 'react';
import { Point } from '@projectstorm/geometry';
import { DiagramEngine } from '../../../src/DiagramEngine';
import { LabelModel } from '../../../src/entities/label/LabelModel';
import { LinkModel } from '../../../src/entities/link/LinkModel';
import { LinkWidget } from '../../../src/entities/link/LinkWidget';
import { PointModel } from '../../../src/entities/link/PointModel';
import { NodeModel } from '../../../src/entities/node/NodeModel';
import { PortModel } from '../../../src/entities/port/PortModel';
import { TestLinkFactory, TestReactFactory } from '../../support/impls/TestLayer';

describe('LinkWidget', () => {
	beforeEach(() => {
		Toolkit.TESTING = true;
		Toolkit.TESTING_UID = 0;
	});
	afterEach(() => {
		Toolkit.TESTING = false;
	});
	test.each([
		{
			name: 'default',
			options: { type: 'test', selected: true },
			labels: []
		},
		{
			name: 'lables',
			options: { type: 'test', selected: false },
			labels: [new LabelModel({ type: 'test' }), new LabelModel({ type: 'test' })]
		},
		{
			name: 'Source Port',
			options: { type: 'test', selected: false },
			labels: [new LabelModel({ type: 'test' }), new LabelModel({ type: 'test' })],
			source: new PortModel({ name: 'source' }),
			target: null
		},
		{
			name: 'Target Port',
			options: { type: 'test', selected: false },
			labels: [new LabelModel({ type: 'test' }), new LabelModel({ type: 'test' })],
			target: new PortModel({ name: 'target' }),
			source: null
		},
		{
			name: 'BOTH Port',
			options: { type: 'test', selected: false },
			labels: [new LabelModel({ type: 'test' }), new LabelModel({ type: 'test' })],
			target: new PortModel({ name: 'target' }),
			source: new PortModel({ name: 'source' })
		},
		{
			name: 'No labels',
			options: { type: 'test', selected: false },
			labels: [],
			target: new PortModel({ name: 'target' }),
			source: new PortModel({ name: 'source' })
		}
	])('Render', async (tc) => {
		const linkModel: LinkModel = new LinkModel(tc.options);
		tc.source ? linkModel.setSourcePort(tc.source) : null;
		tc.source ? tc.source.setParent(new NodeModel({})) : null;
		tc.target ? linkModel.setTargetPort(tc.target) : null;
		tc.target ? tc.target.setParent(new NodeModel({})) : null;
		expect(!!linkModel.getSourcePort()).toBe(!!tc.source);
		tc.labels.forEach((label) => {
			linkModel.addLabel(label);
		});
		const diagramEngine: DiagramEngine = new DiagramEngine({});
		diagramEngine.getLabelFactories().registerFactory(new TestReactFactory('test'));
		diagramEngine.getLinkFactories().registerFactory(new TestLinkFactory('test'));

		const widget = render(<LinkWidget link={linkModel} diagramEngine={diagramEngine}></LinkWidget>, {
			wrapper: ({ children }) => <svg>{children}</svg>
		});

		expect(widget.baseElement).toMatchSnapshot();
		widget.rerender(<LinkWidget link={linkModel} diagramEngine={diagramEngine}></LinkWidget>);
		expect(widget.baseElement).toMatchSnapshot();

		tc.source ? null : linkModel.setSourcePort(new PortModel({ name: 'source' }));
		widget.rerender(<LinkWidget link={linkModel} diagramEngine={diagramEngine}></LinkWidget>);
		expect(widget.baseElement).toMatchSnapshot();

		tc.target ? null : linkModel.setTargetPort(new PortModel({ name: 'tgt' }));
		widget.rerender(<LinkWidget link={linkModel} diagramEngine={diagramEngine}></LinkWidget>);
		expect(widget.baseElement).toMatchSnapshot();
	});

	test('Static', () => {
		const link = new LinkModel({});
		expect(
			LinkWidget.generateLinePath(
				new PointModel({ link, position: new Point(0, 0) }),
				new PointModel({ link, position: new Point(1, 1) })
			)
		).toMatchSnapshot();
	});
});
