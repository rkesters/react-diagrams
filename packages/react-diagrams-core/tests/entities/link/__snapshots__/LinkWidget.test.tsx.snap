// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkWidget Render 1`] = `
<body>
  <div>
    <svg>
      <g
        data-linkid="1"
      >
        <div
          id="linkFactory"
        >
           
        </div>
      </g>
    </svg>
  </div>
</body>
`;

exports[`LinkWidget Render 2`] = `
<body>
  <div>
    <svg>
      <g
        data-linkid="1"
      >
        <div
          id="linkFactory"
        >
           
        </div>
      </g>
    </svg>
  </div>
</body>
`;

exports[`LinkWidget Render 3`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 4`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 5`] = `
<body>
  <div>
    <svg>
      <g
        data-linkid="1"
      >
        <div
          id="linkFactory"
        >
           
        </div>
        <foreignobject
          class="css-1ipt3tw"
        >
          <div
            class="css-157va7r"
          >
            <div
              id="labelFactory"
            >
               
            </div>
          </div>
        </foreignobject>
        <foreignobject
          class="css-1ipt3tw"
        >
          <div
            class="css-157va7r"
          >
            <div
              id="labelFactory"
            >
               
            </div>
          </div>
        </foreignobject>
      </g>
    </svg>
  </div>
</body>
`;

exports[`LinkWidget Render 6`] = `
<body>
  <div>
    <svg>
      <g
        data-linkid="1"
      >
        <div
          id="linkFactory"
        >
           
        </div>
        <foreignobject
          class="css-1ipt3tw"
        >
          <div
            class="css-157va7r"
          >
            <div
              id="labelFactory"
            >
               
            </div>
          </div>
        </foreignobject>
        <foreignobject
          class="css-1ipt3tw"
        >
          <div
            class="css-157va7r"
          >
            <div
              id="labelFactory"
            >
               
            </div>
          </div>
        </foreignobject>
      </g>
    </svg>
  </div>
</body>
`;

exports[`LinkWidget Render 7`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 8`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 9`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 10`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 11`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 12`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 13`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 14`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 15`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 16`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 17`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 18`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 19`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 20`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 21`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 22`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 23`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Render 24`] = `
<body>
  <div>
    <svg />
  </div>
</body>
`;

exports[`LinkWidget Static 1`] = `"M0,0 L 1,1"`;
